File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/CashierSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.db.Cashier;
import eu.pretix.libpretixsync.db.ItemCategory;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;
import io.requery.query.Tuple;
import io.requery.util.CloseableIterator;

public class CashierSyncAdapter extends BaseConditionalSyncAdapter<Cashier, Long> {

    public CashierSyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, PretixApi api, String syncCycleId, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, "__all__", api, syncCycleId, feedback);
    }

    @Override
    public void updateObject(Cashier obj, JSONObject jsonobj) throws JSONException {
        obj.setServer_id(jsonobj.getLong("id"));
        obj.setName(jsonobj.getString("name"));
        obj.setUserid(jsonobj.getString("userid"));
        obj.setLogin_method(jsonobj.getString("login_method"));
        obj.setPin(jsonobj.isNull("pin") ? "" : jsonobj.getString("pin"));
        obj.setOtp_key(jsonobj.isNull("otp_key") ? "" : jsonobj.getString("otp_key"));
        obj.setJson_data(jsonobj.toString());
        obj.setActive(jsonobj.getBoolean("active"));
    }

    @Override
    public CloseableIterator<Cashier> runBatch(List<Long> ids) {
        return store.select(Cashier.class)
                .where(Cashier.SERVER_ID.in(ids))
                .get().iterator();
    }

    @Override
    CloseableIterator<Tuple> getKnownIDsIterator() {
        return store.select(Cashier.SERVER_ID)
                .get().iterator();
    }

    @Override
    String getResourceName() {
        return "cashiers";
    }

    protected String getUrl() {
        return api.organizerResourceUrl("pos/" + getResourceName());
    }

    @Override
    Long getId(JSONObject obj) throws JSONException {
        return obj.getLong("id");
    }

    @Override
    Long getId(Cashier obj) {
        return obj.getServer_id();
    }

    @Override
    Cashier newEmptyObject() {
        return new Cashier();
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/CashierSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import app.cash.sqldelight.db.QueryResult
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.sqldelight.Cashier
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import org.json.JSONObject

class CashierSyncAdapter(
    db: SyncDatabase,
    fileStorage: FileStorage,
    api: PretixApi,
    syncCycleId: String,
    feedback: SyncManager.ProgressFeedback?,
) : BaseConditionalSyncAdapter<Cashier, Long>(
    db = db,
    fileStorage = fileStorage,
    eventSlug = "__all__",
    api = api,
    syncCycleId = syncCycleId,
    feedback = feedback,
) {

    override fun getResourceName(): String = "cashiers"

    override fun getUrl(): String = api.organizerResourceUrl("pos/" + getResourceName())

    override fun getId(obj: Cashier): Long = obj.server_id!!

    override fun getId(obj: JSONObject): Long = obj.getLong("id")

    override fun getJSON(obj: Cashier): JSONObject = JSONObject(obj.json_data)

    override fun queryKnownIDs(): MutableSet<Long> {
        val res = mutableSetOf<Long>()
        db.cashierQueries.selectServerIds().execute { cursor ->
            while (cursor.next().value) {
                val id = cursor.getLong(0)
                    ?: throw RuntimeException("server_id column not available")

                res.add(id)
            }
            QueryResult.Unit
        }

        return res
    }

    override fun insert(jsonobj: JSONObject) {
        db.cashierQueries.insert(
            active = jsonobj.getBoolean("active"),
            json_data = jsonobj.toString(),
            name = jsonobj.getString("name"),
            pin = if (jsonobj.isNull("pin")) "" else jsonobj.getString("pin"),
            server_id = jsonobj.getLong("id"),
            userid = jsonobj.getString("userid"),
        )
    }

    override fun update(obj: Cashier, jsonobj: JSONObject) {
        db.cashierQueries.updateFromJson(
            active = jsonobj.getBoolean("active"),
            json_data = jsonobj.toString(),
            name = jsonobj.getString("name"),
            pin = if (jsonobj.isNull("pin")) "" else jsonobj.getString("pin"),
            userid = jsonobj.getString("userid"),
            id = obj.id,
        )
    }

    override fun delete(key: Long) {
        db.cashierQueries.deleteByServerId(key)
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.cashierQueries.transaction(false, body)
    }

    override fun runBatch(parameterBatch: List<Long>): List<Cashier> =
        db.cashierQueries.selectByServerIdList(parameterBatch).executeAsList()

}
